### ISSUE TITLE
False positive for restrict-plus-operands rule

### ISSUE THREAD

**Issue by chrismbarr** [palantir/tslint#1871]
### Bug Report

- __TSLint version__: `4.0.2`
- __TypeScript version__: `2.1.4`
- __Running TSLint via__: gulp-tslint `7.0.1`

#### TypeScript code being linted

```ts
export const getFilePathInfo = (sourceFile: ts.SourceFile): IFilePathSections => {
    const fullFilePath: string = sourceFile.fileName;
    const lastSlashIdx: number = fullFilePath.lastIndexOf("/");
    const filePath: string = fullFilePath.substr(0, lastSlashIdx + 1);
    const fileName: string = fullFilePath.substr(lastSlashIdx + 1, fullFilePath.length);
    const fileNameParts: string[] = fileName.split(".");
    return {
        fileName: fileName,
        fileNameParts: fileNameParts,
        filePath: filePath
    };
};
```

#### With `tslint.json` configuration:

```json
"rules": {
  "restrict-plus-operands": true,
  //...
}
```

#### Relevant `gulpfile.js` setup
Which is [exactly like the gulp-tslint project recommends](https://github.com/panuhorsmalahti/gulp-tslint/blob/master/gulpfile.js#L8-L22)
```js
const gulp        = require("gulp");
const gulpTslint  = require("gulp-tslint");
const ts          = require("gulp-typescript");
const tslint      = require("tslint");

//TS programs
const tsProject = ts.createProject("./tsconfig.json");
const tslintProgram = tslint.Linter.createProgram("./tsconfig.json", ".");

gulp.task("lint", () => {
    const tsResult = tsProject.src()
        .pipe(gulpTslint({
            formatter: "stylish",
            program: tslintProgram
        }))
        .pipe(gulpTslint.report({
            emitError: false,
            reportLimit: 50
        }))
        .pipe(tsProject());

    return tsResult.js.pipe(gulp.dest(__dirname));
});
```

#### Actual behavior
It reports errors for my addition operations
```
12:53  restrict-plus-operands  Types of values used in '+' operation must match
13:50  restrict-plus-operands  Types of values used in '+' operation must match
```

#### Expected behavior
No errors reported when adding 2 numbers together.  At first I didn't have the explicit `:number` type specified on the variables, so I added it in hopes that it would pick up the types, but no luck!

**Comment by nchen63** [comment]
This may be a ts 2.1 incompatibility issue. This is fixed in `master`
